scio-extra/src/main/scala/com/spotify/scio/extra/Sparkey.scala

```diff
@@ -0,0 +1,166 @@
+    def sparkeySideInput(url: SparkeyUri): SideInput[SparkeyReader] = {
+      val view = self.parallelize(Seq(url)).applyInternal(View.asSingleton())
+      new SparkeySideInput(view)
+    }
+  }
+
+  implicit class SCollectionWithSparkeyWriter[K, V](val self: SCollection[(K, V)])
+                                                   (implicit ev1: K <:< String, ev2: V <:< String) {
+    def asSparkey(uri: SparkeyUri): SCollection[SparkeyUri] = self.transform { in =>
+      in.groupBy(_ => ())
+        .map { case (_, iter) =>
+          val writer = uri.getWriter()
+          val it = iter.iterator
+          while (it.hasNext) {
+            val kv = it.next()
+            writer.put(kv._1.toString, kv._2.toString)
+          }
+          writer.close()
+          uri
+        }
+    }
+
+    def asSparkey: SCollection[SparkeyUri] = {
+      val tempLocation = Try(self.context.optionsAs[DataflowPipelineOptions].getGcpTempLocation)
+        .getOrElse(self.context.options.getTempLocation)
+      this.asSparkey(SparkeyUri(tempLocation))
+    }
+  }
+
+  implicit class SparkeySCollection(val self: SCollection[SparkeyUri]) {
+    def asSparkeySideInput(): SideInput[SparkeyReader] = {
+      val view = self.applyInternal(View.asSingleton())
+      new SparkeySideInput(view)
+    }
+  }
+
+  implicit class SCollectionToSparkeySideInput(val self: SCollection[(String, String)]) {
+    def asSparkeySideInput(): SideInput[SparkeyReader] = self.asSparkey.asSparkeySideInput()
+  }
+
+  implicit class Reader(val self: SparkeyReader) {
+    def get(key: String): Option[String] = Option(self.getAsString(key))
+
+    def toStream(): Stream[(String, String)] = self.iterator().asScala.toStream.map { e =>
+      (e.getKeyAsString, e.getValueAsString)
+    }
+  }
+
+  private[scio] class SparkeySideInput(val view: PCollectionView[SparkeyUri])
+    extends SideInput[SparkeyReader] {
+    override def get[I, O](context: DoFn[I, O]#ProcessContext): SparkeyReader =
+      SparkeyUri(context.sideInput(view).path).getReader()
+  }
+}
+
+trait SparkeyUri extends Serializable {
+  val path: String
+  def getWriter(): WrappedSparkeyWriter
+  def getReader(): SparkeyReader
+}
+
+object SparkeyUri {
+  def apply(path: String): SparkeyUri =
+    if (new URI(path).getScheme == "gs") new GcsSparkeyUri(path) else new LocalSparkeyUri(path)
+}
+
+private case class LocalSparkeyUri(path: String) extends SparkeyUri {
+  override def getReader(): SparkeyReader = JSparkey.open(new File(path + ".spi"))
+
+  override def getWriter(): WrappedSparkeyWriter = new LocalSparkeyWriter(path)
+}
+
+private case class GcsSparkeyUri(path: String) extends SparkeyUri {
+  override def getReader(): SparkeyReader = {
+    val gcs = new GcsUtilFactory().create(PipelineOptionsFactory.create())
+    val localTmpDir = Files.createTempDir()
+    // Copy .spi and .spl to local files
+    for (ext <- Seq(".spi", ".spl")) {
+      val src = gcs.open(GcsPath.fromUri(s"$path/sparkey$ext"))
```

Comment:
Just saw that you're using `Files.createTempDir()`, but this will download one copy per thread to a different location. See `DistCache` for a stable unique copying mechanism.